planner/optimize.go

```diff
@@ -264,6 +293,22 @@ func optimize(ctx context.Context, sctx sessionctx.Context, node ast.Node, is in
 		return nil, nil, 0, err
 	}
 
+	if !sctx.GetSessionVars().InRestrictedSQL {
+		readOnly, err := sctx.GetSessionVars().GlobalVarsAccessor.GetGlobalSysVar(variable.TiDBRestrictedReadOnly)
```

Comment:
I'm not sure if this will impact performance, since there is a lot of mess that goes into the code path of `GetGlobalSysVar`. It is also potentially more difficult to unit test.

When I was working on sem, I used an atomic global value for enable/disable:
https://github.com/pingcap/tidb/blob/209334951cd920e5408f67fd60a0b542507526ee/util/sem/sem.go#L69-L87

I copied the design from the implementation of collations.

You can write code to detect the change of `tidb_restricted_read_only` and update your atomic value here https://github.com/pingcap/tidb/blob/209334951cd920e5408f67fd60a0b542507526ee/domain/sysvar_cache.go#L148-L206 (unfortunately there is not a more elegant way to detect a sysvar change from other servers, the comment provides details why.)